In [0]:
import gym
import numpy as np


In [0]:
env = gym.make('Taxi-v3')
#always gamma .9
gamma = 0.90
#find correct value
epsilon = .05
#find correct value
alpha = .6
episodes = 1500000 

Q = np.zeros((env.observation_space.n, env.action_space.n))


In [0]:
def choose_action(state):
  action=0
  
  if np.random.random() < epsilon:
    action = env.action_space.sample()
    
  else:
    action = np.argmax(Q[state, :])
  #print(action)
  return action

In [0]:
def learn(state, state2, reward, action):
    predict = Q[state, action]
    target = reward + gamma * np.max(Q[state2, :])
    Q[state, action] = Q[state, action] + alpha * (target - predict)

In [65]:
for episode in range(episodes):
    state = env.reset()
    t = 0
    
    while True:
        #env.render()

        action = choose_action(state)  

        state2, reward, done, info = env.step(action)  

        learn(state, state2, reward, action)

        state = state2

        t += 1
       
        if done:
            break

        
#Q(462, 4) = −11.374
#Q(398, 3) = 4.348
#Q(253, 0) = −0.585
#Q(377, 1) = 9.683
#Q(83, 5) = −13.996
print(Q[462,4])
print(Q[398, 3])
print(Q[253, 0])
print(Q[377, 1])
print(Q[83, 5])

-11.374402515013
4.348907000000002
-0.5856821172999982
9.683000000000002
-13.996845490100029


In [66]:
print(Q[46,3])
print(Q[38,0])
print(Q[442,0])
print(Q[12,5])
print(Q[321,0])
print(Q[359,3])
print(Q[53,1])
print(Q[172,1])
print(Q[496,1])
print(Q[47,3])


-1.5271139055699985
7.714700000000002
-1.5271139055699985
-13.440939433444477
-3.8232660371605287
5.943229389351537
-1.5271139055699985
-2.3744025150129984
2.9140163000000023
1.6226146700000021


Q(444,4)

-13.44
 
Question 2
10 / 10 pts
Q(373,0)

4.348
 
Question 3
10 / 10 pts
Q(131,1)

-3.136
 
Question 4
10 / 10 pts
Q(126,0)

-3.823
 
IncorrectQuestion 5
0 / 10 pts
Q(359,3)

5.923
 
Question 6
10 / 10 pts
Q(12,5)

-13.44
 
Question 7
10 / 10 pts
Q(136,3)

14.299
 
Question 8
10 / 10 pts
Q(66,2)

0.46
 
Question 9
10 / 10 pts
Q(343,1)

-2.374
 
Question 10
10 / 10 pts
Q(252,4)

-10.527
